internal/backend/schedule/scheduleworkflow.go

```diff
@@ -0,0 +1,150 @@
+		return fmt.Errorf("schedule wf: worker start: %w", err)
+	}
+	swf.Z.Info("registered workflow and worker", zap.String("workflow", wf.SchedulerWorkflow), zap.String("worker", wf.SchedulerWorkerID))
+	return nil
+}
+
+func (swf *SchedulerWorkflow) getSessionData(ctx context.Context, triggerID sdktypes.TriggerID) ([]wf.SessionData, error) {
+	z := swf.Z.With(zap.String("trigger_id", triggerID.String()))
+
+	if !triggerID.IsValid() {
+		z.Debug("invalid trigger")
+		return nil, fmt.Errorf("trigger: %w", sdkerrors.ErrNotFound)
+	}
+
+	trigger, err := swf.Services.Triggers.Get(ctx, triggerID)
+	if err != nil {
+		return nil, fmt.Errorf("get trigger: %w", err)
+	}
+
+	if trigger.EventType() != sdktypes.SchedulerEventTriggerType { // sanity
+		z.Error("unsupported trigger type, expected `scheduler`", zap.String("trigger_type", trigger.EventType()))
+		return nil, fmt.Errorf("unexpected trigger type. Need `scheduler`, got %q", trigger.EventType())
+	}
+
+	envID := trigger.EnvID()
+	if !envID.IsValid() {
+		z.Debug("invalid environment")
+		return nil, fmt.Errorf("environment: %w", sdkerrors.ErrNotFound)
+	}
+
+	depFilter := sdkservices.ListDeploymentsFilter{State: sdktypes.DeploymentStateActive, EnvID: envID}
+	deployments, err := swf.Services.Deployments.List(ctx, depFilter)
+	if err != nil {
+		z.Panic("could not fetch active deployments", zap.Error(err))
+	}
+
+	if len(deployments) == 0 {
+		z.Debug("no deployments")
+		return nil, nil
+	}
+
+	cl := trigger.CodeLocation()
+	var sds []wf.SessionData
+	for _, dep := range deployments {
+		sds = append(sds, wf.SessionData{Deployment: dep, CodeLocation: cl, Trigger: trigger})
+		z.Debug("deployment found", zap.String("deployment_id", dep.ID().String()))
+	}
+
+	return sds, nil
+}
+
+func (swf *SchedulerWorkflow) newScheduleTickEvent(ctx context.Context, triggerID sdktypes.TriggerID) sdktypes.Event {
+	z := swf.Z.With(zap.String("trigger_id", triggerID.String()))
+
+	event := kittehs.Must1(sdktypes.EventFromProto(
+		&sdktypes.EventPB{
+			EventType:    sdktypes.SchedulerTickEventType,
+			Memo:         map[string]string{"trigger_id": triggerID.String()},
+			ConnectionId: fixtures.BuiltinSchedulerConnectionID.String(),
+		}))
+
+	eventID, err := swf.Services.Events.Save(ctx, event) // create event
+	if err != nil {
+		z.Panic("failed creating schedule tick event")
+	}
+
+	// REVIEW: re-fetching newly created event is costly (db). Right now we are recreating event without CreatedAt.
+	// other solution is to cause events.SaveEvent to return created event and not just ID
+	event = kittehs.Must1(sdktypes.EventFromProto(&sdktypes.EventPB{EventId: eventID.String(), EventType: sdktypes.SchedulerTickEventType}))
+
+	swf.CreateEventRecord(ctx, eventID, sdktypes.EventStateProcessing) // add `processing` event record
+	return event
+}
+
+func (swf *SchedulerWorkflow) scheduleWorkflowInternal(wfctx workflow.Context, ctx context.Context, triggerID sdktypes.TriggerID, tickEvent *sdktypes.Event) error {
```

Comment:
this is used only in scheduleWorkflow, why did you pull this out?

Reply:
I wanted to create internal function that does all work and logs with zap.Logger and external function that creates event ,updates state and logs with logger. The way how this is structured in dispatcher is much less readable IMO
I could rename this function of coz. `scheduleWorkflowInternal` does all relevant schedule work

Reply:
resolving?

Reply:
need to think of a better name

Reply:
ScheduleWorkflowEntry and ScheduleWorkflow?
ScheduleWorkflow and ScheduleWorkflowRunSessions?

Reply:
scheduleSessions